File: src/main/java/org/raku/comma/annotation/fix/RemoveUnusedRoutineFix.java
package org.raku.comma.annotation.fix;

import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.codeInspection.util.IntentionFamilyName;
import com.intellij.codeInspection.util.IntentionName;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiWhiteSpace;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.IncorrectOperationException;
import org.raku.comma.psi.RakuRoutineDecl;
import org.raku.comma.psi.RakuStatement;
import org.raku.comma.utils.RakuPsiUtil;
import org.jetbrains.annotations.NotNull;

public class RemoveUnusedRoutineFix implements IntentionAction {
    @Override
    public @IntentionName @NotNull String getText() {
        return "Safe delete routine";
    }

    @Override
    public @NotNull @IntentionFamilyName String getFamilyName() {
        return "Raku";
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        return true;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
        PsiElement el = file.findElementAt(editor.getCaretModel().getOffset());
        if (el == null)
            return;
        PsiElement decl = PsiTreeUtil.getParentOfType(el, RakuRoutineDecl.class);
        if (decl == null)
            return;
        RakuPsiUtil.deleteElementDocComments(PsiTreeUtil.getParentOfType(decl, RakuStatement.class));
        PsiElement maybeWS = decl.getParent().getNextSibling();
        if (maybeWS instanceof PsiWhiteSpace)
            maybeWS.delete();
        decl.getParent().delete();
    }

    @Override
    public boolean startInWriteAction() {
        return true;
    }
}

File: src/main/java/org/raku/comma/inspection/fixes/RemoveUnusedRoutineFix.kt
package org.raku.comma.inspection.fixes

import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.util.IntentionName
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiTreeUtil
import org.raku.comma.psi.RakuRoutineDecl
import org.raku.comma.psi.RakuStatement
import org.raku.comma.utils.RakuPsiUtil

class RemoveUnusedRoutineFix(private val name: String): LocalQuickFix {
    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
// TODO: Should we bother with this caretModel stuff?
//        val file: PsiFile = descriptor.psiElement.containingFile
//        val editor = PsiEditorUtil.findEditor(el)
//        val el: PsiElement = file.findElementAt(editor.getCaretModel().getOffset()) ?: return
//        val el = descriptor.psiElement
//        val decl: PsiElement = PsiTreeUtil.getParentOfType(el, RakuRoutineDecl::class.java) ?: return
        val declaration = descriptor.psiElement as RakuRoutineDecl
        RakuPsiUtil.deleteElementDocComments(PsiTreeUtil.getParentOfType(declaration, RakuStatement::class.java))
        val maybeWS = declaration.parent.nextSibling
        if (maybeWS is PsiWhiteSpace) maybeWS.delete()
        declaration.parent.delete()
    }

    override fun getName(): @IntentionName String { return "Safe delete routine '%s'".format(name) }
    override fun getFamilyName(): String { return "Safe delete routine" }
}